src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,106 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+flb_sds_t replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    flb_sds_t result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        flb_sds_destroy(result);
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag)
+{
+    int i = 0;
+    char tag_idx[9];
+    char datestamp[40];
+    flb_sds_t tag_token;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t time_format = NULL;
+    struct tm *gmt;
+    flb_sds_t tag_delimiter = ".";
+    char tag_descriptor[10] = "$TAG";
+    char open[2] = "[";
+    char close[2] = "]";
+
+    s3_key = flb_sds_create_size((strlen(format) + 1) * sizeof(char));
+    if (!s3_key) {
+        flb_errno();
+        flb_sds_destroy(s3_key);
+        return NULL;
+    }
+    strcpy(s3_key, format);
+    /* Finds all occurences of $TAG[*] and replaces then with the right token from tag.*/
+    tag_token = strtok(tag, tag_delimiter);
+
+    while(tag_token != NULL) {
+        char tmp[20] = "";
+        sprintf(tag_idx, "%d", i);
+        strcat(tmp, tag_descriptor);
+        strcat(tmp, open);
+        strcat(tmp, tag_idx);
+        strcat(tmp, close);
+
+        s3_key = replace_uri_tokens(s3_key, tmp, tag_token);
```

Comment:
your replace_uri_tokens function will allocate a new string, so you are over-writing the previous value of s3_key here- that's a memory leak.